### PR TITLE
common/BackTrace: add operator<<

### PR DESCRIPTION
**Pull request by tchaikov**
could be handy when using along with dout.

Signed-off-by: Kefu Chai tchaikov@gmail.com

### CODE REVIEW

**Comment by dachary** [comment]
jenkins test this please (no logs)

**Comment by tchaikov** [comment]
@badone could you help take a look, thanks!

**Comment by yuriw** [comment]
test this pls

**Comment by yuriw** [comment]
test this please

File: src/test/common/test_back_trace.cc [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+TEST(BackTrace, Basic) {
+  std::string bt = foo();
+  std::vector<std::string> lines;
+  boost::split(lines, bt, boost::is_any_of("\n"));
+  const unsigned lineno = 1;
+  ASSERT_GT(lines.size(), lineno);
+  std::regex e("^ 1: "
+               "\\(foo.*\\)\\s"
+               "\\[0x[[:xdigit:]]+\\]$");
+  EXPECT_TRUE(std::regex_match(lines[lineno], e));
```
Comment (badone) [inline src/test/common/test_back_trace.cc]: At runtime there appears to be an issue with this code as it is throwing an exception, "regex_error". Need to wrap it in a try/catch and output std::regex_error::what to see what it's complaining about.

Comment (badone) [inline src/test/common/test_back_trace.cc]: Tried to catch this with https://github.com/ceph/ceph/pull/12933 but all test passed :(

**Comment by badone** [comment]
Test this please

File: src/test/common/test_back_trace.cc (hunk 1 above)
Comment (badone) [inline src/test/common/test_back_trace.cc]: Oops, forgot to return an error condition. Trying again.

**Comment by tchaikov** [comment]
@badone turns out the regex implemented in libstdc++ prior to 4.9 was buggy, see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53631. so let use `boost::regex` at this moment. thanks for looking at this failure!

**Comment by badone** [comment]
The test is still failing with boost::regex

**Comment by badone** [comment]
Test this please

**Comment by badone** [comment]
Looks good now :)

**Comment by tchaikov** [comment]
yeah, i figured out that we should add `-fno-inline` last night.

**Comment by yuriw** [comment]
@tchaikov 

this was tested along with others:

http://pulpito.ceph.com/yuriw-2017-01-13_21:05:46-rados-wip-yuri-testing2_2017_1_13-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-01-14_16:29:03-rados-wip-yuri-testing2_2017_1_13---basic-smithi/

(see http://tracker.ceph.com/issues/18478)

centos run
http://pulpito.front.sepia.ceph.com/yuriw-2017-01-14_16:34:35-rados-wip-yuri-testing2_2017_1_13-distro-basic-vps/

One unit test failed `the following tests FAILED: 82 - unittest_back_trace (Failed)` and @liewegas requested @tchaikov to see if this is related and can be fixed

Here is the backtrace:

`yuriw@smithi098:~/wip2/build$  ctest -R unittest_back_trace -V
UpdateCTestConfiguration  from :/home/yuriw/wip2/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/yuriw/wip2/build/DartConfiguration.tcl
Test project /home/yuriw/wip2/build
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 82
    Start 82: unittest_back_trace

82: Test command: /home/yuriw/wip2/build/bin/unittest_back_trace
82: Environment variables: 
82:  CEPH_ROOT=/home/yuriw/wip2
82:  CEPH_BIN=/home/yuriw/wip2/build/bin
82:  CEPH_LIB=/home/yuriw/wip2/build/lib
82:  CEPH_BUILD_DIR=/home/yuriw/wip2/build
82:  LD_LIBRARY_PATH=/home/yuriw/wip2/build/lib
82:  PATH=/home/yuriw/wip2/build/bin:/home/yuriw/wip2/src:/home/yuriw/bin:/home/yuriw/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
82:  PYTHONPATH=/home/yuriw/wip2/build/lib/cython_modules/lib.2:/home/yuriw/wip2/src/pybind
82:  CEPH_BUILD_VIRTUALENV=/tmp
82: Test timeout computed to be: 3600
82: Running main() from gmock_main.cc
82: [==========] Running 1 test from 1 test case.
82: [----------] Global test environment set-up.
82: [----------] 1 test from BackTrace
82: [ RUN      ] BackTrace.Basic
82: /home/yuriw/wip2/src/test/common/test_back_trace.cc:36: Failure
82: Value of: std::regex_match(lines[lineno], e)
82:   Actual: false
82: Expected: true
82: [  FAILED  ] BackTrace.Basic (1 ms)
82: [----------] 1 test from BackTrace (1 ms total)
82: 
82: [----------] Global test environment tear-down
82: [==========] 1 test from 1 test case ran. (1 ms total)
82: [  PASSED  ] 0 tests.
82: [  FAILED  ] 1 test, listed below:
82: [  FAILED  ] BackTrace.Basic
82: 
82:  1 FAILED TEST
1/1 Test #82: unittest_back_trace ..............***Failed    0.02 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.04 sec

The following tests FAILED:
	 82 - unittest_back_trace (Failed)
Errors while running CTest
`

**Comment by yuriw** [comment]
@tchaikov pls merge or assign back to me if needed, meanwhile un-tagged

**Comment by tchaikov** [comment]
@yuriw the unit test was fixed. and the code have been tested in the qa run was not changed. merging.

**Comment by wjwithagen** [comment]
On 17-1-2017 04:12, Kefu Chai wrote:
> @yuriw <https://github.com/yuriw> the unit test was fixed. and the code
> have been tested in the qa run was not changed. merging.

Hi Kefu,

On FreeBSD with Clang with the new code I still get:
> build/bin/unittest_back_trace
Running main() from gmock_main.cc
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from BackTrace
[ RUN      ] BackTrace.Basic
/usr/srcs/Ceph/work/ceph/src/test/common/test_back_trace.cc:40: Failure
Value of: boost::regex_match(lines[lineno], e)
  Actual: false
Expected: true
[  FAILED  ] BackTrace.Basic (7 ms)
[----------] 1 test from BackTrace (7 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (7 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] BackTrace.Basic

 1 FAILED TEST

And that is independant on -fno-inline....

--WjW

**Pull request merged**